### PR TITLE
Allow running arbitrary scripts via `autokey-run` and phrase macros

### PR DESCRIPTION
**Pull request by BlueDrink9**
If the script argument given to `autokey-run` or a phrase macro like `<script name>` is an absolute path (with `~` expanded to `$HOME`), the script at that path will be run as if an autokey script, instead of searching for a script with that description.

This allows scripts to be run without their containing folder being manually added in the gui.

Also updates macro argument parsing, to allow quoted arguments with spaces in them (something which the wiki currently notes as a known limitation.)

Also makes a number of fixes to validation of hotkey creation. Sorry about not extracting this, it was a bit on-the-fly when I was testing a script.

### CODE REVIEW

File: lib/autokey/macro.py [hunk 1]
```diff
@@ -95,23 +111,23 @@ def _can_process(self, token):
...
-        ret = {}
-
-        if len(l) > 1:
-            for arg in l[1:]:
-                key, val = arg.split('=', 1)
-                ret[key] = val
-
-        for k, v in self.ARGS:
-            if k not in ret:
-                raise Exception("Missing mandatory argument '{}' for macro '{}'".format(k, self.ID))
```
Comment (luziferius) [inline lib/autokey/macro.py:95]: Since this PR is against `develop`, changing the Exception type to ValueError is ok. This is something I wanted to do anyways, so it’s nice to see it done.

File: lib/autokey/macro.py [hunk 2]
```diff
@@ -30,6 +31,21 @@ def on_triggered(self):
...
+    extracted = [p.split('>')[0] for p in s.split('<') if '>' in p]
+    if len(extracted) == 0:
+        return s
+    else:
+        return ''.join(extracted)
+
+def split_key_val(s):
+    # Split as if a shell argument.
+    # Splits at spaces, but preserves spaces within quotes.
+    pairs = shlex.split(s)
```
Comment (luziferius) [inline lib/autokey/macro.py:46]: Using shlex is an interesting idea.
When I started to work on this, I threw the biggest hammer on the problem I knew by the time, by building a full-fledged RE-based tokenizer and parser. My goal was to have a fully bug- and corner-case free parser. Unfortunately, there was no real end result. Just for reference, that’s the diff of my incomplete and stale attempt: https://github.com/autokey/autokey/compare/develop...luziferius:improve_phrase_macro_parsing

So let’s see how your much shorter solution copes with various corner cases.

File: tests/test_macro.py [hunk 3]
```diff
@@ -13,13 +13,53 @@
...
+    # first-run logic.
+    with patch("autokey.model.Phrase.persist"), patch("autokey.model.Folder.persist"),\
+         patch("autokey.configmanager.configmanager.ConfigManager.load_global_config",
+               new=(lambda self: self.folders.append(test_folder))):
+        engine = Engine(ConfigManager(MagicMock()), MagicMock(spec=PhraseRunner))
+        engine.configManager.config_altered(False)
+
+    return engine, test_folder
+
+def test_arg_parse():
```
Comment (luziferius) [inline tests/test_macro.py:48]: With this kind of structural regularity in the test inputs, you can parametrize the test and save the typing/copy&pasting the same assert statement over and over again.
just add 3 arguments to the test case function like `test_input, expected, error_msg` and put
the `@pytest.mark.parametrize` decorator in the front of the definition.
Like
```python
@pytest.mark.parametrize("test_input, expected, error_msg", [
    ("<script name='test name' args='long arg with spaces and ='>", {"name": "test name", "args": "long arg with spaces and ="}, "Macro arg can't contain equals"),
…
])
def test_arg_parse(test_input, expected, error_msg):
    engine, folder = create_engine()
    macro = ScriptMacro(engine)
    assert_that(macro._get_args(test_input), is_(equal_to(expected)),
                error_msg)
```
Then adding a new case is just adding a new tuple with test data.

File: lib/autokey/service.py [hunk 4]
```diff
@@ -532,4 +532,5 @@ def run_subscript(self, script):
 
     def run_subscript_path(self, path):
         scope = self.scope.copy()
+        scope["__file__"] = script.path
```
Comment (luziferius) [inline lib/autokey/service.py]: Oh yeah, Seems I forgot that location. Thanks for fixing it

File: lib/autokey/service.py [hunk 5]
```diff
@@ -491,6 +498,20 @@ def execute(self, script: model.Script, buffer=''):
...
 
+    @threaded
+    def execute_path(self, path: pathlib.Path, buffer=''):
+        logger.debug("Script runner executing: {}".format(path))
+        scope = self.scope.copy()
+        # scope["store"] = script.store
+        # Overwrite __file__ to contain the path to the user script instead of the path to this service.py file.
+        scope["__file__"] = path.resolve()
+        try:
+            exec(open(path).read(), scope)
```
Comment (luziferius) [inline lib/autokey/service.py]: This `open(path).read()` leaks an opened file handle, which is never closed.
`path.read_text()` automatically closes the file after reading the content, as indicated by
https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text

File: lib/autokey/macro.py (hunk 2 above)
Comment (BlueDrink9) [reply lib/autokey/macro.py:46]: Can you tell me what sorts of corner cases you were referring to? Then I can add tests, at least.

Comment (BlueDrink9) [reply lib/autokey/macro.py:46]: Sorry, I hadn't realised you had already put so much work in. I just came across needing spaces in args, and assumed there'd be a library to parse strings for arguments.

Comment (BlueDrink9) [reply lib/autokey/macro.py:46]: Yours looks like it will probably be more sophisticated. What stage is it at? May be worth replacing my quick change with yours when it is done. Mine doesn't handle nested macros, and possibly never can.

File: tests/test_macro.py (hunk 3 above)
Comment (BlueDrink9) [reply tests/test_macro.py:48]: Thank you! I saw that in one of the other test files, but didn't understand how it was used.

File: lib/autokey/service.py (hunk 5 above)
Comment (BlueDrink9) [reply lib/autokey/service.py]: Interesting! I note that it was only added in 3.5. I don't know what you consider a minimal supported python 3 to be.

I'll change it. Teaches me to blindly copy the SO solution.

Comment (BlueDrink9) [reply lib/autokey/service.py]: Oh, actually it wasn't SO, it was an equivalent function somewhere else in autokey. So I'll change it twice.

File: lib/autokey/macro.py (hunk 2 above)
Comment (luziferius) [reply lib/autokey/macro.py:46]: > Sorry, I hadn't realised you had already put so much work in. I just came across needing spaces in args, and assumed there'd be a library to parse strings for arguments.

No Problem! That branch lives inside my personal fork (see the diff link) and I consider it stale at the moment. It is incomplete. There’s no way I’d assume you look around my stuff to look for bits and pieces. So don’t worry about it.


> Can you tell me what sorts of corner cases you were referring to? Then I can add tests, at least.

Let’s see if I can still pull some from under my hat. I had mostly nesting broken macro tokens and similar stuff on my mind.


- `<script name="<script name="foo", args="None">", args="None">` This nests two valid macro tokens, if you consider `<script name="foo", args="None">` to be a valid script name (which it is!). Which one to execute is to decide…
- `<script <script name="myscript" args="valid_token">>`
- `<script name=<file <date format="%Y \"> some text <script (broken) <script name="myscript" args="valid_token">> args="<script name=`

Comment (BlueDrink9) [reply lib/autokey/macro.py:46]: Ofc I wouldn't look at your bits, but neovim for example requires PRs discussing what you plan to do, to prevent duplication of work/unwanted work. At the least, we could set up some sort of communication system for what we are working on.

Comment (BlueDrink9) [reply lib/autokey/macro.py:46]: Ah, they look quite involved. The second one currently isn't valid because the outer macro requires 'args', still.

TBH my method still doesn't handle nested macros anyway, let alone broken ones. I can look at implementing it, but it's lower priority unless you make it a requirement for merge.

Comment (luziferius) [reply lib/autokey/macro.py:46]: > Ah, they look quite involved. The second one currently isn't valid because the outer macro requires 'args', still.

In the second point, note that the outer `<script>`-thing has no `name` parameter. Here, the inner token is valid.

> TBH my method still doesn't handle nested macros anyway, let alone broken ones. I can look at implementing it, but it's lower priority unless you make it a requirement for merge.

That won’t be a requirement, so it’s fine as it is.

File: tests/test_macro.py (hunk 3 above)
Comment (luziferius) [reply tests/test_macro.py:48]: In those, I went a step further. You can pass any iterable, like a Generator function instead of a list into `@pytest.mark.parametrize()`.

So I wrote functions named like `generate_test_cases_for_<insert_test_name_here>()` and placed them in front of the test function. Then I used some nested helper functions inside the generator functions to cut down typing redundant bits and used those to `yield` one test case per line.

This is quite advanced but makes it super easy to add additional test cases to existing tests.

File: lib/autokey/service.py (hunk 5 above)
Comment (luziferius) [reply lib/autokey/service.py]: > Oh, actually it wasn't SO, it was an equivalent function somewhere else in autokey. So I'll change it twice.

Huh? OK then, TIL. Go ahead ;)

> Interesting! I note that it was only added in 3.5. I don't know what you consider a minimal supported python 3 to be.

Huh? (The second…) The Readme used to be stating that Python 3.5 is the minimal supported Python version.

Comment (BlueDrink9) [reply lib/autokey/service.py]: Oops, sorry. Should have checked the readme.

I'll see if I can find that function, I might me misremembering.

Comment (luziferius) [reply lib/autokey/service.py]: No no no. Sorry from my side. It seems that this bit vanished when the dependency information was de-duplicated. There used to be a slightly outdated section in the README and in the Wiki. Both got merged into the Wiki. But it seems that the minimal Python version was not properly carried to the Wiki page.

Edit: I’ve updated the section, here: https://github.com/autokey/autokey/wiki/Installing#Dependencies

Comment (BlueDrink9) [reply lib/autokey/service.py]: I take it back, it wasn't already there. Normally there was no need to read the code in a script, I added it as part of arbitrary scripts. But I think there are two instances, I'll change both.

Statement about SO stands.

Comment (BlueDrink9) [reply lib/autokey/service.py]: Also, regarding the wiki: As I make changes, they change relevancy bits of wiki documentation. But it seems the wiki can only apply to the current latest release. Any suggestion for storing up notes on what to update in the wiki? Maybe a file somewhere with notes

Comment (luziferius) [reply lib/autokey/service.py]: Oh, And about:

> Teaches me to blindly copy the SO solution.

You are sure that the license of the copied SO solution was GPLv3 compatible? :smiling_imp: 

(This, plus possible issues in the actual code are the two major obstacles when looking up stuff there…)



> Also, regarding the wiki: As I make changes, they change relevancy bits of wiki documentation. […]

I think, handling this with some additional notes at the relevant bits may be sufficient. For example, noting at which version functions were added and changed. For example, somewhat like this: https://documen.tician.de/pyopencl/runtime_queue.html#pyopencl.Event.from_int_ptr

That documentation has statements like these, to indicate changes:

> New in version 2013.2.
> 
> Changed in version 2016.1: retain added.

Comment (BlueDrink9) [reply lib/autokey/service.py]: Ok. So if I make a change in my branch, I hope it will eventually become a PR, which will eventually be merged into dev, which will become `0.96`. What stage do I say in the wiki "Changed in `0.96`"?

File: tests/test_macro.py (hunk 3 above)
Comment (BlueDrink9) [reply tests/test_macro.py:48]: I'll push this change to `macro` branch, pull #340

File: lib/autokey/service.py (hunk 5 above)
Comment (luziferius) [reply lib/autokey/service.py]: I’d say some time prior to releasing the next version. The changelog has to be written anyways, so that time frame looks like it may be a good idea. I’m not totally sure either.
The best way may be actually placing it into the docstrings and automatically generate the relevant pages. (Related to #330)

Comment (BlueDrink9) [reply lib/autokey/service.py]: I wholeheartedly agree with the last part, but for macros that could be problematic because they aren't associated with a particular function to create a docstring. Unless we
just chuck the docstring in `macro.py`

File: tests/test_macro.py (hunk 3 above)
Comment (BlueDrink9) [reply tests/test_macro.py:48]: Any suggestions for how to parametrize `test_engine.py`'s `create_phrase`? Since it needs a `Folder` object created, but doesn't always use it?

File: lib/autokey/service.py (hunk 5 above)
Comment (BlueDrink9) [reply lib/autokey/service.py]: This has been fixed.

File: tests/test_macro.py (hunk 3 above)
Comment (luziferius) [reply tests/test_macro.py:48]: Partition the test cases by the used parameters and the expected result (positive/success, negative/raisedException). Each partition of the tests gets it’s own function. Then parameters can be used to raise the number of tests by testing multiple valid/invalid inputs.

Comment (BlueDrink9) [reply tests/test_macro.py:48]: But how do you pass an object arg instead of a string without making it a global arg?

Comment (BlueDrink9) [reply tests/test_macro.py]: But how do you pass an object arg instead of a string without making it a global arg?

**Comment by BlueDrink9** [comment]
I think there is still a bug with running arbitrary scripts from phrase macros. Once I fix it, this will be ready to merge.

**Comment by BlueDrink9** [comment]
I've done some manual checks, and I think this is ready to be merged unless there's more you want to do. Merge before #340.

**Comment by josephj11** [comment]
We're going to need some documentation on how to use this when it is released. Sounds pretty powerful.

**Comment by BlueDrink9** [comment]
I agree Joseph. The problem is the wiki isn't version controlled, which is annoying. But I've in-script documented the api function, at least.

File: tests/test_macro.py (hunk 3 above)
Comment (luziferius) [reply tests/test_macro.py:48]: > But how do you pass an object arg instead of a string without making it a global arg?

Sorry, it seems that I missed your last question.

Normally, you can pass anything as an argument to the test function.
But If you create class objects of any AutoKey class inside the test parameter list, any Exceptions caused during the object creation won’t be associated to the failing test case. 

Instead, those Exceptions will be risen during the test collection phase (because the test case generators are run beforehand). And Exceptions during the collection phase cause more things to fail.

I’m not quite sure, but it may even kill the whole test suite with an "Errors occurred during the collection phase." error. At least, it fails the whole test.

If we start to use `pyfakefs` to test file system interactions, the `Path` instances must be created inside the `test_function()`, otherwise it will point to the outside world. There may be similar issues with other packages.

That’s why I’m restricting myself to use built-in types or simple structs of built-in types only.
